File: src/test/java/by/architecture_map/belarus/service/impl/ImageServiceImplTest.java
package by.architecture_map.belarus.service.impl;

import by.architecture_map.belarus.entity.Image;
import by.architecture_map.belarus.exception.NotFoundException;
import by.architecture_map.belarus.repository.jpa.ImageRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class ImageServiceImplTest {

    @Mock
    private ImageRepository imageRepository;

    @InjectMocks
    private ImageServiceImpl imageService;

    @Test
    void whenFind_thenReturnImage() {
        int id = 1;
        Image image = new Image();
        image.setId(id);
        when(imageRepository.findById(id)).thenReturn(Optional.of(image));

        Image result = imageService.find(id);

        assertThat(result).isNotNull();
        assertThat(result.getId()).isEqualTo(id);
        verify(imageRepository).findById(id);
    }

    @Test
    void whenNotFound_thenThrowNotFoundException() {
        int id = 2;
        when(imageRepository.findById(id)).thenReturn(Optional.empty());

        NotFoundException exception = assertThrows(NotFoundException.class, () -> imageService.find(id));

        assertThat(exception.getMessage().contains(String.valueOf(id)));
        verify(imageRepository).findById(id);
    }
}

File: src/test/kotlin/by/architecturemap/belarus/service/impl/ImageServiceImplTest.kt
package by.architecturemap.belarus.service.impl

import by.architecturemap.belarus.entity.Image
import by.architecturemap.belarus.exception.NotFoundException
import by.architecturemap.belarus.repository.jpa.ImageRepository
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.assertThrows
import java.util.Optional

class ImageServiceImplTest {

    private val imageRepository: ImageRepository = mockk()
    private val imageService = ImageServiceImpl(imageRepository)

    private val id = 1

    @Test
    fun givenImageId_whenFindImage_thenReturnImage() {

        //arrange
        val expected = Image(url = "https://test.com")
        every { imageRepository.findById(id) } returns Optional.of(expected)

        //act
        val actual = imageService.find(id)

        //assert
        assertEquals(expected, actual)
    }

    @Test
    fun givenImageId_whenFindImageAndImageNotFound_thenThrowNotFoundException() {

        //arrange
        every { imageRepository.findById(id) } returns Optional.empty()

        //act & assert
        assertThrows<NotFoundException> { imageService.find(id) }
        verify(exactly = 1) { imageRepository.findById(id) }
    }
}
